group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ServerSideStickyRangeAssignor.java

```diff
@@ -0,0 +1,268 @@
+                    }
+                }
+                // Number of partitions left to reach the minRequiredQuota
+                int remaining = minRequiredQuota - currentAssignmentSize;
+
+                // There are 3 cases w.r.t value of remaining
+                // 1) remaining < 0 this means that the consumer has more than the min required amount.
+                // It could have an extra partition, so we check for that.
+                if (remaining < 0 && numConsumersWithExtraPartition > 0) {
+                    // In order to remain as sticky as possible, since the order of members can be different, we want the consumers that already had extra
+                    // partitions to retain them if it's still required, instead of assigning the extras to the first few consumers directly.
+                    // Ex:- If two consumers out of 3 are supposed to get an extra partition and currently 1 of them already has the extra, we want this consumer
+                    // to retain it first and later if we have partitions left they will be assigned to the first few consumers and the unfilled map is updated
+                    numConsumersWithExtraPartition--;
+                    // Since we already added the minimumRequiredQuota of partitions in the previous step (until minReq - 1), we just need to
+                    // add the extra partition which will be present at the index right after min quota is satisfied.
+                    putSet(assignedStickyPartitionsPerTopic, topicId, currentAssignmentListForTopic.get(minRequiredQuota));
+                    membersWithNewAssignmentPerTopic.computeIfAbsent(memberId, k -> new HashMap<>()).computeIfAbsent(topicId, k -> new HashSet<>()).add(currentAssignmentListForTopic.get(minRequiredQuota));
+                } else {
+                    // 3) If remaining = 0 it has min req partitions but there is scope for getting an extra partition later on, so it is a potentialUnfilledConsumer.
+                    // 4) If remaining > 0 it doesn't even have the min required partitions, and it definitely is unfilled so it should be added to potentialUnfilledConsumers.
+                    Pair<String, Integer> newPair = new Pair<>(memberId, remaining);
+                    potentiallyUnfilledConsumers.add(newPair);
+                }
+            }
+
+            // Step 3
+            // Iterate through potentially unfilled consumers and if remaining > 0 after considering the extra partitions assignment, add to the unfilled list.
+            for (Pair<String, Integer> pair : potentiallyUnfilledConsumers) {
+                String memberId = pair.getFirst();
+                Integer remaining = pair.getSecond();
+                if (numConsumersWithExtraPartition > 0) {
+                    remaining++;
```

Comment:
can you help me understand why we're incrementing?

remaining is the number of partitions remaining to meet the min required quota so if we're giving one of the extra partitions to this member, my intuition tells me that remaining should decremented here.

Reply:
remaining = minRequiredQuota - completedQuota

Reply:
ah okay. so we're increasing the completed by 1